src/operator/semijoin.rs

```diff
@@ -0,0 +1,609 @@
+        semijoin_func: F,
+    ) -> Stream<Circuit<()>, Out>
+    where
+        F: Fn(&Pairs::Key, &Pairs::Val) -> Out::Key + Clone + 'static,
+        Pairs: IndexedZSet + DeepSizeOf + Send,
+        Pairs::Key: Hash + Ord + Clone + DeepSizeOf,
+        Pairs::Val: Ord + Clone,
+        Pairs::R: MulByRef + DeepSizeOf,
+        Keys: ZSet<Key = Pairs::Key, R = Pairs::R> + DeepSizeOf + Send,
+        Out: ZSet<R = Pairs::R>,
+    {
+        let (pairs, keys) = (self.shard(), keys.shard());
+        pairs
+            .integrate_trace()
+            .delay_trace()
+            .semijoin_stream_core_inner(&keys, semijoin_func.clone())
+            .plus(&pairs.semijoin_stream_core_inner(&keys.integrate_trace(), semijoin_func))
+    }
+}
+
+impl<S, Pairs> Stream<Circuit<S>, Pairs>
+where
+    S: Clone + 'static,
+{
+    // TODO: Derive `TS` type from circuit.
+    /// Incremental join two streams of batches.
+    ///
+    /// Given streams `self` and `other` of batches that represent changes to
+    /// relations `A` and `B` respectively, computes a stream of changes to
+    /// `A <> B` (where `<>` is the join operator):
+    ///
+    /// # Type arguments
+    ///
+    /// * `Pairs` - batch type in the first input stream.
+    /// * `Keys` - batch type in the second input stream.
+    /// * `Out` - output Z-set type.
+    pub fn semijoin<Time, F, Keys, Out>(
+        &self,
+        keys: &Stream<Circuit<S>, Keys>,
+        semijoin_func: F,
+    ) -> Stream<Circuit<S>, Out>
+    where
+        Time: Timestamp + DeepSizeOf,
+        F: Fn(&Pairs::Key, &Pairs::Val) -> Out::Key + 'static,
+        Pairs: IndexedZSet + Send,
+        Pairs::Key: Hash + Ord + DeepSizeOf + Clone,
+        Pairs::Val: Ord + DeepSizeOf + Clone,
+        Pairs::R: MulByRef + Default + DeepSizeOf,
+        Keys: ZSet<Key = Pairs::Key, R = Pairs::R> + Send,
+        Out: ZSet<R = Pairs::R>,
+        Out::Key: Clone,
+        Out::Batcher: DeepSizeOf,
+    {
+        let key_trace = keys
+            .shard()
+            .trace::<OrdValSpine<Pairs::Key, Keys::Val, Time, Pairs::R>>();
+
+        // FIXME: Is this correct?
+        self.circuit().add_binary_operator(
+            SemiJoinTrace::new(semijoin_func),
+            &self.shard(),
+            &key_trace,
+        )
+    }
+}
+
+pub struct SemiJoinTrace<F, Pairs, Keys, Out>
+where
+    Keys: TraceReader,
+    Out: ZSet,
+{
+    semijoin_func: F,
+    // TODO: not needed once timekeeping is handled by the circuit.
+    time: Keys::Time,
+    // Future update batches computed ahead of time, indexed by time
+    // when each batch should be output.
+    output_batchers: HashMap<Keys::Time, Out::Batcher>,
+    // True if empty input batch was received at the current clock cycle.
+    empty_input: bool,
+    // True if empty output was produced at the current clock cycle.
+    empty_output: bool,
+    _types: PhantomData<(Pairs, Keys, Out)>,
+}
+
+impl<F, Pairs, Keys, Out> SemiJoinTrace<F, Pairs, Keys, Out>
```

Comment:
Likewise, this should be as efficient as `JoinTrace`, so do we really want to maintain a mostly identical implementation?